BGV/CKKS: support scale management (#1459)
google/heir: docs/content/en/docs/Design/management.md

would be

1. After multiplication: `input -> (mult -> ms) -> (mult -> ms) -> output`

1. Before multiplication: `input -> (mult) -> (ms -> mult) -> (ms -> output)`

1. Before multiplication (including the first multiplication):
   `input -> (ms -> mult) -> (ms -> mult) -> (ms -> output)`

The first strategy is from the BGV paper, the second and third strategies are
from OpenFHE, which correspond to the `FLEXIBLEAUTO` mode and `FLEXIBLEAUTOEXT`
mode, respectively.

The first strategy is conceptually simpler, yet other policies have the
advantage of smaller noise growth. In latter policies, at the beginning of each
level, the noise level would be the level of multiplication instead of modulus
switching, so the error incurred by other operations (relinearization/rotation)
in the same level could be hidden.

Note that, as multiplication has two operands, the actual circuit for the latter
two policies is `mult(ms(ct0), ms(ct1))`, whereas in the first policy the
circuit is `ms(mult(ct0, ct1))`.

The third policy has one more modulus switching than others, so it will need one
more modulus.

There are also other insertion strategy like inserting it dynamically based on

> current noise (see HElib) or lazy rescaling. Those are not implemented.

Author: It becomes `lazy modulus switching` now. It is an optimization.

computation happens in `Zt`, then by dividing of `qi` the result message would
be `[m / qi]_t`.

Note that when `qi = 1 (mod t)`, the result message is the same as the original
message. However, in GHS variant such condition does not always hold, so we
would call the `[q^{-1}]_t` part as the _scale_ of the message and compiler
needs to record and manage it during computation. And when decrypting such scale
must be removed to obtain the original message.

Note that, for messages `m0` and `m1` of different scale `a` and `b`, we could
not add them directly because `[a * m0 + b * m1]_t` is often not the expected
result for user. Instead we need to adjust the scale of one message to match
another so `[b * m0 + b * m1]_t = [b * (m0 + m1)]_t`. Such adjustment could be
done by multiplying `m0` with a constant `[b / a]_t`. This adjustment is not for
free as it will also amplify the noise.

As one may expect, different modulus switching insertion strategy would affect
the scale on the message. For `m0` with scale `a` and `input1` with scale `b`,
the result scale would be

1. After multiplication: `[ab / qi]_t`.

1. Before multiplication: `[a / qi * b / qi]_t` = `[ab / (qi^2)]_t`.

> This is messy enough. To ease the burden, we can impose additional requirement:
> Mandate a constant scale `Delta_i` for each level. This is called
> _level-specific scaling factor_. With this in mind, addition within one level
> can happen without caring about the scale, and the scale of the next level is
> also the same.

Reviewer: ```suggestion
This is messy enough. To ease the burden, we can impose additional requirement:
mandate a constant scale `Delta_i` for all ciphertexts at level `i`. 
This is called the _level-specific scaling factor_.
With this in mind, addition within one level
can happen without caring about the scale.
```

1. After multiplication: `Delta_{i-1} = [Delta_i^2 / qi]_t`

1. Before multiplication: `Delta_{i-1} = [Delta_i^2 / (qi^2)]_t`

### BGV: Cross Level Operation

> With the level-specific scaling factor, one may wonder how could addition and
> multiplication of ciphertext of different levels happen as they have different
> scaling factor. This can be solved by adjusting the level and scale of the
> ciphertext of higher level.

Reviewer: ```suggestion
With the level-specific scaling factor, one may wonder how to perform addition and
multiplication of ciphertexts on different levels. This can be done by adjusting
the level and scale of the ciphertext at the higher level.
```

The level can be easily adjusted by dropping the extra limbs, and scale can be
adjusted by multiplying a constant, but because multiplying a constant will
incur additional noise, the procedure becomes the following:

1. Assume the level and scale of two ciphertexts are `l1` and `l2`, `s1` and
   `s2` respectively. WLOG assume `l1 > l2`.

1. Drop `l1 - l2 - 1` limbs for the first ciphertext to make it at level
   `l2 + 1`, if those extra limbs exist.

1. Adjust scale from `s1` to `s2 * q_{l2 + 1}` by multiplying
   `[s2 * q_{l2 + 1} / s1]_t` for the first ciphertext.

> 1. Modulusing switching from `l2 + 1` to `l2`, the scale becomes `s2` for the

Reviewer: ```suggestion
1. Modulus switch from `l2 + 1` to `l2`, producing scale `s2` for the
```

   first ciphertext.

### BGV: Implementation in HEIR

In HEIR the different modulus switching policy is controlled by the pass option
for `--secret-insert-mgmt-bgv`. The pass defaults to the "Before Multiplication"
policy. If user wants other policy, the `after-mul` or
`before-mul-include-first-mul` option may be used. The `mlir-to-bgv` pipeline
option `modulus-switch-before-first-mul` corresponds to the latter option.

The `secret-insert-mgmt` pass is also responsible for managing cross-level

> operations. However, as the scheme parameters has not been generated, the

Reviewer: ```suggestion
operations. However, as the scheme parameters are not generated at this point, the
```

concrete scale could not be instantiated so some placeholder operations are
inserted.

> After the modulus switching policy applied, user could use `generate-param-bgv`
> pass to generate scheme parameters. Optionally, user could skip this pass by
> manually providing scheme parameter as an attribute at module level.

Reviewer: ```suggestion
After the modulus switching policy is applied, the `generate-param-bgv`
pass generates scheme parameters. Optionally, user could skip this pass by
manually providing scheme parameter as an attribute at module level.
```

Then `populate-scale-bgv` comes into play by using the scheme parameters to
instantiate concrete scale, and turn those placeholder operations into concrete
multiplication operation.

## CKKS

CKKS is a leveled scheme where each level has a modulus `qi`. The level is
numbered from `0` to `L` where `L` is the input level and `0` is the output

> level. CKKS ciphertext contains a scaled message $\\Delta m$ where $\\Delta$

Reviewer: Though I am not an expert here, I had believed that the scale in CKKS was also partly used for message precision, in the sense of a fixed-point numeric scale (and given the infinite input domain for CKKS encoding). But for BGV it feels mainly like an artifact of the GHS optimization. Are there any meaningful differences in how these two kinds of "scale" are handled? I'm guessing based on this document that the answer is not for the purposes of `mgmt`, but maybe there is something about choosing an input encoding scale that is left unsaid.
